src/main/java/com/sap/cloudfoundry/client/facade/oauth2/OAuthClient.java

```diff
@@ -48,67 +51,62 @@ public void clear() {
-
-        ResourceOwnerPasswordAccessTokenProvider provider = new ResourceOwnerPasswordAccessTokenProvider();
-
-        return provider.refreshAccessToken(resource, token.getRefreshToken(), request);
-    }
-
     public TokenProvider getTokenProvider() {
         return new OAuthTokenProvider(this);
     }
 
-    private AccessTokenRequest createAccessTokenRequest(String username, String password) {
-        return new DefaultAccessTokenRequest();
+    private boolean shouldRefreshToken() {
+        return credentials.isRefreshable() && token.getOAuth2AccessToken()
+                                                   .getExpiresAt()
+                                                   .isBefore(Instant.now()
+                                                                    .plus(50, ChronoUnit.SECONDS));
     }
 
-    private OAuth2ProtectedResourceDetails getResourceDetails(String username, String password, String clientId, String clientSecret) {
-        ResourceOwnerPasswordResourceDetails resource = new ResourceOwnerPasswordResourceDetails();
-        resource.setUsername(username);
-        resource.setPassword(password);
+    protected OAuth2AccessTokenWithAdditionalInfo createToken() {
+        MultiValueMap<String, String> formData = new LinkedMultiValueMap<>();
+        formData.add("grant_type", "password");
+        formData.add("client_id", credentials.getClientId());
+        formData.add("client_secret", credentials.getClientSecret());
+        formData.add("username", credentials.getEmail());
+        formData.add("password", credentials.getPassword());
+        formData.add("response_type", "token");
+        Oauth2AccessTokenResponse oauth2AccessTokenResponse = fetchOauth2AccessToken(formData);
+        return tokenFactory.createToken(oauth2AccessTokenResponse);
+    }
 
-        resource.setClientId(clientId);
-        resource.setClientSecret(clientSecret);
-        resource.setId(clientId);
-        resource.setClientAuthenticationScheme(AuthenticationScheme.header);
-        resource.setAccessTokenUri(authorizationUrl + "/oauth/token");
+    private Oauth2AccessTokenResponse fetchOauth2AccessToken(MultiValueMap<String, String> formData) {
+        try {
+            return webClient.post()
+                            .uri(authorizationUrl + "/oauth/token")
+                            .header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_FORM_URLENCODED_VALUE)
+                            .body(BodyInserters.fromFormData(formData))
+                            .retrieve()
+                            .bodyToFlux(Oauth2AccessTokenResponse.class)
+                            .blockFirst();
+        } catch (WebClientResponseException e) {
+            throw new ResponseStatusException(e.getStatusCode(), e.getMessage(), e);
+        }
+    }
 
-        return resource;
+    private TokenFactory getTokenFactory() {
```

Comment:
If this method is private and therefore it won't be overridden in tests, why not inline it?